cpp/src/neighbors/detail/cagra/cagra_search.cuh

```diff
@@ -108,6 +111,114 @@ void search_main_core(raft::resources const& res,
+ * @tparam IdxT type of database vector indices
+ * @tparam internal_IdxT during search we map IdxT to internal_IdxT, this way we do not need
+ * separate kernels for int/uint.
+ *
+ * @param[in] handle
+ * @param[in] params configure the search
+ * @param[in] strided_dataset CAGRA strided dataset
+ * @param[in] metric distance type
+ * @param[in] queries a device matrix view to a row-major matrix [n_queries, index->dim()]
+ * @param[out] neighbors a device matrix view to the indices of the neighbors in the source dataset
+ * [n_queries, k]
+ * @param[out] distances a device matrix view to the distances to the selected neighbors [n_queries,
+ * k]
+ * @param[in] sample_filter a device filter function that greenlights samples for a given query
+ *
+ * @return true If the brute force search was applied successfully.
+ * @return false If the brute force search was not applied.
+ */
+template <typename T,
+          typename InternalIdxT,
+          typename CagraSampleFilterT,
+          typename IdxT      = uint32_t,
+          typename DistanceT = float>
+bool search_using_brute_force(
+  raft::resources const& res,
+  const search_params& params,
+  const strided_dataset<T, IdxT>& strided_dataset,
+  cuvs::distance::DistanceType metric,
+  raft::device_matrix_view<const T, int64_t, raft::row_major> queries,
+  raft::device_matrix_view<InternalIdxT, int64_t, raft::row_major> neighbors,
+  raft::device_matrix_view<DistanceT, int64_t, raft::row_major> distances,
+  CagraSampleFilterT& sample_filter)
+{
+  if (params.threshold_to_bf >= 1.0) { return false; };
+
+  auto n_queries = queries.extent(0);
+  auto n_dataset = strided_dataset.n_rows();
+
+  auto bitset_filter_view = sample_filter.bitset_view_;
+  auto sparsity           = bitset_filter_view.sparsity(res);
+
+  if (sparsity < params.threshold_to_bf) { return false; }
+
+  // TODO: Support host dataset in `brute_force::build`
+  RAFT_LOG_DEBUG("CAGRA is switching to brute force with sparsity:%f", sparsity);
+  using bitmap_view_t = cuvs::core::bitmap_view<const uint32_t, int64_t>;
+
+  auto stream            = raft::resource::get_cuda_stream(res);
+  auto bitmap_n_elements = bitmap_view_t::eval_n_elements(bitset_filter_view.size() * n_queries);
+
+  rmm::device_uvector<uint32_t> raw_bitmap(bitmap_n_elements, stream);
+  rmm::device_uvector<int64_t> raw_neighbors(neighbors.size(), stream);
+
+  bitset_filter_view.repeat(res, n_queries, raw_bitmap.data());
+
+  auto brute_force_filter = bitmap_view_t(raw_bitmap.data(), n_queries, n_dataset);
+
+  auto brute_force_neighbors = raft::make_device_matrix_view<int64_t, int64_t, raft::row_major>(
+    raw_neighbors.data(), neighbors.extent(0), neighbors.extent(1));
+  auto brute_force_dataset = raft::make_device_matrix_view<const T, int64_t, raft::row_major>(
+    strided_dataset.view().data_handle(), strided_dataset.n_rows(), strided_dataset.stride());
+
+  auto brute_force_idx = cuvs::neighbors::brute_force::build(res, brute_force_dataset, metric);
```

Comment:
This is being called each and every time a user performs a search? There's overhead in this call, and this should cache off the built brute-force index because for many common distances this computes a set of norms.